### DISCUSSION TITLE
[Issue]: NPM commands give error: 11: exec: node: not found

### DISCUSSION THREAD

**Discussion by tragicrich** [coreybutler/nvm-windows#1154] [Support]
### What happened?

I have been battling an issue for months where I am unable to get NVM / Node to work on my desktop (works fine on my other machine).

Error from any repo running an npm command: 11: exec: node: not found

Via powershell I ran _nvm debug_ "no problems detected"
I can run nvm -v, node -v, npm -v and get proper version results

C:\Program Files\nodejs is a symlink created by the exe that shows the current node files
My environment variables have %NVM_HOME% and %NVM_SYMLINK% with proper paths, and no other conflicting paths for nodejs

I have uninstalled node / nvm multiple times and reinstalled and continue to get the same error.
I have run _nvm on_
I have run Powershell / CMD with Administrative Priveleges
I have scoured Stack Overflow and tried any and all recommendations from both accepted answers and other suggested comments

I am unable to determine why npm commands are not working. Everything returns a correct result until I run an actual command.

The differences from my other machine are:
1. My desktop is Win 64bit
2. My development files are on D:\

I am losing my mind on this and would like to be able to develop on my main machine again. Any help is greatly appreciated.

### What did you expect to happen?

When running an npm command, my desktop should use the chosen version installed via NVM and should not give a "node not found" error.

### Version

1.1.11 or newer (Default)

### Which version of Windows?

Windows 11+

### Which locale?

English (Default)

### Which shell are you running NVM4W in?

Command Prompt, PowerShell

### User Permissions?

Administrative Privileges, Elevated

### Is Developer Mode enabled?

Yes

### Relevant log/console output

```Shell
11: exec: node: not found
```

### Debug Output

```Shell
Administrator: Windows PowerShell
NVM4W Version:      1.1.11
NVM4W Path:         C:\Users\<user>\AppData\Roaming\nvm\nvm.exe
NVM4W Settings:     C:\Users\<user>\AppData\Roaming\nvm\settings.txt
NVM_HOME:           C:\Users\<user>\AppData\Roaming\nvm
NVM_SYMLINK:        C:\Program Files\nodejs
Node Installations: C:\Users\<user>\AppData\Roaming\nvm

Active Node.js Version: v19.9.0

No problems detected.
```

**Comment by coreybutler** [comment]
I'm sorry you're going through this. That sounds frustrating.

The first thing I'd suggest is reviewing any `.npmrc` files in your repos. This can change how npm behaves and where it looks for the node binary. Dotfiles can often be outdated, especially if they're not committed to git repos.

In the "unlikely, but still possible" category, are you installing NVM4WW with a forked version? This is very important because there was recently a fork that added junction support over the default symlink behavior. This is known to fail on hard drives that are not used for the installation (i.e. your `D:\`).

Additionally, did `C:\Program Files\nodejs` exist before the installation, or was it created manually? If it existed before the installation, the symlink would never be correctly generated when running `nvm use 19.9.0` You can easily resolve this by deleting `C:\Program Files\nodejs` and re-running `nvm use 19.9.0`... just make sure `C:\Program Files\nodejs` is a symlink and not an old version of Node (otherwise you'll nuke the old version). This would be pretty unlikely though, unless a prior install of v19.9.0 existed when NVM4W was installed.

Hopefully one of these things will lead you to an answer. If not, let me know.

Since I don't think this is a bug, I'm going to move it to a discussion.

**Comment by tragicrich** [comment]
I very much appreciate the quick reply / feedback.

1. I don't see any .npmrc files in the repos I am testing
2. Unlikely. I just downloaded the latest version here: https://github.com/coreybutler/nvm-windows/releases. Now showing NVM4W Version:      1.1.12 No problems Detected
3. Program Files/nodejs was removed prior to installation / upgrade and was recreated by nvm

Still receiving the same error

**Reply by coreybutler** [reply]
Within your project, can you run the npm command using the fully qualified path?

For example, `C:\Users\corey\AppData\Roaming\nvm\v19.9.0` is the physical location of v19.9.0 on my machine:

<img width="563" alt="image" src="https://github.com/coreybutler/nvm-windows/assets/770982/26ddecc3-d44d-4379-b609-4f5b11240f4a">

Within the project directory, I would run `C:\Users\corey\AppData\Roaming\nvm\v19.9.0\npm.cmd run mytask`. If you can run this directly, it suggests a PATH issue.

Additionally, remember `.npmrc` files can exist outside of project directories, as in parent directories or in the global `node_modules` directory (or even the node root).

Finally, have you tried explicitly installing the 64 bit version, i.e. `nvm install 19.9.0 64`. I don't think that would create a problem with npm, but I'm trying to rule out oddities.

**Reply by tragicrich** [reply]
Thank you for the follow up Corey!

Running the npm.cmd with full path gives me the same error
![image](https://github.com/user-attachments/assets/9bb8fcb7-567c-4b1f-adde-e5d76016e10c)

Trying to install 19.9.0 64 says the version is already installed. Uninstalling and reinstalling with the 64 flag and full pathname outputs the same error:
/mnt/d/drivey/node_modules/.bin/vue-cli-service: 11: exec: node: not found

I searched C: and D: for .npmrc, most were within the nvm folder but I did find an instance at C:\Users\richr I removed the "prefix" line which had the path prefix=C:\Users\richr\AppData\Roaming\nvm\node_modules. Restarted my terminal, executed the full path command and it still gives me the error. I found another instance in AppData, but it only had a path for Python. There were two files npmrcmd and npmrc5 in one of my project folders on D: and I removed them.

**Reply by coreybutler** [reply]
Hmm, this might be something else specific to your project. The vue app is looking for `/mnt/d/drivey/node_modules/.bin` for the vue service (as you noted). It's the `/mnt` part that has me guessing, and the fact it is in unix path format, not Windows path formatting.

In your package.json, add an npm script like this:

```js
{
  ...
  scripts: {
    "nvmtest": "echo test"
  }
}
```

Then run `npm run nvmtest`. This should output `test` in the console, without the error.

_If you don't get the error_, it means the problem is with the npm script, not with node/npm/nvm4w. I suspect this is the case given the output I'm seeing. I suspect the vue-cli-service is looking for node in a specific hard-coded/non-standard location, or perhaps it is treating it like a symlink/not following (I don't know how that tool works, those are just common problems in many CLI tools).

_If you still get the error_, then I'm at a loss, especially since `npm -v` is returning properly. I'd backup `npm.cmd` and start editing it with breakpoints (i.e. some simple echo commands) to troubleshoot further.

**Reply by tragicrich** [reply]
Thank you!

So this is happening with all projects on my D: (vue, angular, gatsby, etc). But your note made me realize this is an issue with WSL2 somehow

I tried unmounting my D drive, but that isn't working. It seems WSL automatically mounts all my drives. I ended up deleting all of the files I could on D (though I can't remove some system files) and recloning two projects. But I am still encountering errors on install

![image](https://github.com/user-attachments/assets/606c5ca0-da01-4141-8150-b266f49b8b5a)
![image](https://github.com/user-attachments/assets/c1cc0831-8544-4404-bd48-252c0a4d5c97)

Do I have to do something special to get WSL and NVM to play nice? To be clear, I am not working in WSL. But it is active on my system.

**Reply by coreybutler** [reply]
That's tricky. I don't know if there is much that can be done. Ultimately, NVM4W just sets the paths to the Node binaries (and therefore npm), then gets out of the way.

What we've collectively confirmed is npm _is executing_. That's as far as NVM4W goes, so it has done its job. If there is a trick/solution, it's going to come from the npm executable.

The key info to note in your latest screenshot is the `npm ERR! command bash -c node ./download-chromedriver.js` (and similar for node-gyp). The fact npm is trying to execute bash means it expects bash to be present, which won't be the case on Windows when you're not in a WSL VM (unless you're using some cygwin magic).

I wish I could offer more of a solution, but I think this boils down to an npm problem, not an nvm4w one.

**Reply by tragicrich** [reply]
Understood. Thanks for all of your help Corey!

**Reply by coreybutler** [reply]
@tragicrich While this won't help you right now, the NVM4W successor ([Runtime](https://github.com/coreybutler/nvm-windows/wiki/Runtime)) will shim npm and other common dependency managers. The goal is to do "pass-thru" commands (i.e. not disturbing native functionality), but the shim allows us to inspect errors from tools like npm when they occur. Runtime will also be cross-platform. This combination will make it possible to identify and possibly accommodate environments like the one you're working in.

**Comment by tragicrich** [comment]
For anyone else encountering issues with NVM and WSL2, the solution is stupidly easy 🤦‍♂️

You simply need to do all of your work moving forward in WSL after it is enabled. This includes install NVM in the subsystem and running node commands from within WSL.
https://learn.microsoft.com/en-us/windows/dev-environment/javascript/nodejs-on-wsl
